### PR TITLE
Lsig

### PR DESCRIPTION
**Pull request by barnjamin**
wip

### CODE REVIEW

File: beaker/precompile.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+        self.program = lsig.program
+
+        self.binary = None
+        self.addr = None
+        self.map = None
+
+        self.template_values: list[TemplateValue] = []
+
+        # Replace the program text with 0 value
```
Comment (deanstef) [inline beaker/precompile.py:56]: Why with 0? We should replace it with the arg value, shouldn't we?

Comment (barnjamin) [reply beaker/precompile.py:56]: we replace with 0 prior to assembly, later the actual value is patched into the binary directly after encoding it properly

File: beaker/precompile.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    def address(self) -> Expr:
+        assert self.binary is not None
+        assert len(self.template_values) == 0
+
+        return Addr(self.addr)
+
+    def signer(self) -> LogicSigTransactionSigner:
+        return LogicSigTransactionSigner(LogicSigAccount(self.binary))
+
+    def populate_template(self, *args) -> bytes:
```
Comment (deanstef) [inline beaker/precompile.py:90]: I see, here you populate the Lsig template with the given args. Very nice.

File: beaker/precompile.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                if type(arg) is not int:
+                    raise Exception("no")
+                curr_val = py_encode_uvarint(arg)
+
+            populated_binary[tv.pc + offset : tv.pc + offset + 1] = curr_val
+            offset += len(curr_val) - 1
+
+        return bytes(populated_binary)
+
+    def template_signer(self, *args) -> LogicSigTransactionSigner:
```
Comment (deanstef) [inline beaker/precompile.py:115]: Does it means that we can have `signer` != `template_signer`?

Comment (barnjamin) [reply beaker/precompile.py:115]: I'm not sure what this mean?

Comment (deanstef) [reply beaker/precompile.py:115]: Oh wait, I think I got confused here.`signer` is the Lsig with 0 values and the `template_signer` the one populated with the actual values. Right?

**Pull request merged**